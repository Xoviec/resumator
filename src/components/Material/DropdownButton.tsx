import { FunctionComponent, useState } from "react";
import * as React from "react";
import { Button, ButtonProps } from "@mui/material";
import { Menu, MenuItem } from "@mui/material";

interface DropdownButtonProps extends Omit<ButtonProps, "onClick"> {
  actions: string[];
  onClick: (action: string) => void;
  testIdPrefix?: string;
}

export const DropdownButton: FunctionComponent<DropdownButtonProps> = ({
  actions,
  children,
  onClick,
  testIdPrefix = "",
  ...props
}) => {
  const [anchorEl, setAnchorEl] = useState<Element | null>(null);

  /**
   * Set the target for the menu to bind to.
   */
  const handleClick = (event: React.MouseEvent) => {
    setAnchorEl(event.currentTarget);
  };

  /**
   * Remove the target for the menu.
   */
  const handleClose = () => {
    setAnchorEl(null);
  };

  /**
   * Handle a menu item click.
   */
  const handleAction = (action: string) => {
    handleClose();
    onClick(action);
  };

  return (
    <>
      {/* Trigger button with added properties from this component. */}
      <Button onClick={handleClick} {...props}>
        {children}
      </Button>
      {/* The menu with possible actions. */}
      <Menu
        keepMounted
        anchorEl={anchorEl}
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        {actions.map((action) => (
          <MenuItem
            key={action}
            onClick={() => handleAction(action)}
            data-testid={
              testIdPrefix
                ? `${testIdPrefix}-${action.toLowerCase()}`
                : action.toLowerCase()
            }
          >
            {action}
          </MenuItem>
        ))}
      </Menu>
    </>
  );
};
