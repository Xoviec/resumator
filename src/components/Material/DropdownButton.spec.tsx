import { fireEvent, render } from "@testing-library/react";
import { DropdownButton } from "./DropdownButton";

describe("DropdownButton", () => {
  it("should render correctly", () => {
    render(
      <DropdownButton
        variant="contained"
        actions={["Action1", "Action2"]}
        startIcon={<div />}
        onClick={jest.fn()}
        color="primary"
      />
    );

    // @mui/Menu renders a portal
    expect(document.body).toMatchSnapshot();
  });

  it("should open the menu when clicked", () => {
    const { getByText } = render(
      <DropdownButton
        variant="contained"
        actions={["Action1", "Action2"]}
        startIcon={<div />}
        onClick={jest.fn()}
        color="primary"
      >
        Main Button
      </DropdownButton>
    );

    fireEvent.click(getByText("Main Button"));
    expect(getByText("Action1")).toBeVisible();
    expect(getByText("Action2")).toBeVisible();
  });

  it("should call onClick when an action is clicked", () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <DropdownButton
        variant="contained"
        actions={["Action1", "Action2"]}
        startIcon={<div />}
        onClick={onClick}
        color="primary"
      >
        Main Button
      </DropdownButton>
    );

    fireEvent.click(getByText("Action1"));
    expect(onClick).toHaveBeenCalledWith("Action1");
  });
});
