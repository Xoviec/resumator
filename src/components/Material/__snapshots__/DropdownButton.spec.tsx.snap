// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownButton should render correctly 1`] = `
<body>
  <div>
    <button
      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <div />
      </span>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    aria-hidden="true"
    class="MuiModal-root MuiModal-hidden MuiPopover-root MuiMenu-root css-6f6ff7-MuiModal-root-MuiPopover-root-MuiMenu-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiBackdrop-invisible css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 0; visibility: hidden;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiMenu-paper MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 MuiPopover-paper css-1poimk-MuiPaper-root-MuiMenu-paper-MuiPaper-root-MuiPopover-paper"
      style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
        role="menu"
        tabindex="-1"
      >
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="0"
        >
          Action1
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          role="menuitem"
          tabindex="-1"
        >
          Action2
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
